Source/WebCore/rendering/MarkedText.cpp

```diff
@@ -113,19 +114,39 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
         auto& parentRenderer = *renderer.parent();
         auto& parentStyle = parentRenderer.style();
         if (auto highlightRegister = renderer.document().highlightRegisterIfExists()) {
-            for (auto& highlight : highlightRegister->map()) {
-                auto renderStyle = parentRenderer.getUncachedPseudoStyle({ PseudoId::Highlight, highlight.key }, &parentStyle);
+            for (auto& highlightName : highlightRegister->vector()) {
+                auto renderStyle = parentRenderer.getUncachedPseudoStyle({ PseudoId::Highlight, highlightName }, &parentStyle);
                 if (!renderStyle)
                     continue;
                 if (renderStyle->textDecorationsInEffect().isEmpty() && phase == PaintPhase::Decoration)
                     continue;
-                for (auto& rangeData : highlight.value->rangesData()) {
+                for (auto& rangeData : highlightRegister->map().get(highlightName)->rangesData()) {
                     if (!highlightData.setRenderRange(rangeData))
                         continue;
 
                     auto [highlightStart, highlightEnd] = highlightData.rangeForTextBox(renderer, selectableRange);
-                    if (highlightStart < highlightEnd)
-                        markedTexts.append({ highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlight.key });
+
+                    if (highlightStart < highlightEnd) {
+                        long currentPriority = highlightRegister->map().get(highlightName)->priority();
+                        // if we can just append it to the end, do that instead
+                        if (markedTexts.isEmpty()
+                            || markedTexts.last().priority <= currentPriority) {
+                            markedTexts.append({ highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlightName, currentPriority });
+                        } else {
+                            // find the first correct place to insert highlight
+                            bool inserted = false;
+                            for (int index = markedTexts.size()-1; index >= 0; index--) {
+                                if (markedTexts[index].priority <= currentPriority) {
+                                    markedTexts.insert(index + 1, { highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlightName, currentPriority });
+                                    inserted = true;
+                                    break;
+                                }
+                            }
+                            // insert at front of vector if lower priority than everything in markedTexts
+                            if (!inserted)
+                                markedTexts.insert(0, { highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlightName, currentPriority });
```

Comment:
Do you know how often we're inserting at the front?

Reply:
In the worst case, if everything in the vector is in descending priority, we would have to keep inserting at the front for every element (which is not great).